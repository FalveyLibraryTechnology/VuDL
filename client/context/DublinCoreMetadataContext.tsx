import React, { createContext, useContext, useReducer } from "react";
import PropTypes from "prop-types";

export interface DublinCoreMetadata {
    currentDublinCore: Record<string, Array<string>>;
    // We need to be careful about how we set keys for elements in our control group; if we reuse
    // keys after changing components, then when we add or remove values in the list, React may
    // inappropriately reuse existing elements instead of correctly redrawing the modified list.
    // By maintaining a counter of the number of times impactful changes are made in the list, we
    // can ensure that keys are always unique, and that elements are redrawn when necessary.
    // This feels like a hack, and a better solution would be welcomed!
    keyCounter: Record<string, number>;
}

const dublinCoreContextParams: DublinCoreMetadata = { currentDublinCore: {}, keyCounter: {} };
const DublinCoreMetadataContext = createContext({});

/**
 * Update the shared states of react components.
 */
const reducerMapping: Record<string, string> = {
    SET_CURRENT_DUBLIN_CORE: "currentDublinCore",
};

const incrementKeyCounter = (keyCounter: Record<string, number>, field: string): Record<string, number> => {
    const current = keyCounter[field] ?? 0;
    return {
        ...keyCounter,
        [field]: current + 1,
    };
};

const addAbove = (dublinCore: Record<string, Array<string>>, field: string, index: number, value: string): Record<string, Array<string>> => {
    const current = (dublinCore[field] ?? []).slice(0);
    current.splice(index, 0, value);
    return {
        ...dublinCore,
        [field]: current,
    }
};

const addBelow = (dublinCore: Record<string, Array<string>>, field: string, index: number, value: string): Record<string, Array<string>> => {
    const current = (dublinCore[field] ?? []).slice(0);
    current.splice(index + 1, 0, value);
    return {
        ...dublinCore,
        [field]: current,
    }
};

const deleteValue = (dublinCore: Record<string, Array<string>>, field: string, index: number): Record<string, Array<string>> => {
    const current = (dublinCore[field] ?? []).slice(0);
    current.splice(index, 1);
    return {
        ...dublinCore,
        [field]: current,
    }
};

const replaceValue = (dublinCore: Record<string, Array<string>>, field: string, index: number, value: string): Record<string, Array<string>> => {
    const current = (dublinCore[field] ?? []).slice(0);
    current[index] = value;
    return {
        ...dublinCore,
        [field]: current,
    }
};

const dublinCoreMetadataReducer = (state: DublinCoreMetadata, { type, payload }: { type: string, payload: unknown }) => {
    if (type === "ADD_VALUE_ABOVE") {
        const { field, index, value } = payload as { field: string, index: number, value: string };
        return {
            ...state,
            keyCounter: incrementKeyCounter(state.keyCounter, field),
            currentDublinCore: addAbove(state.currentDublinCore, field, index, value),
        }
    } else if (type === "ADD_VALUE_BELOW") {
        const { field, index, value } = payload as { field: string, index: number, value: string };
        return {
            ...state,
            keyCounter: incrementKeyCounter(state.keyCounter, field),
            currentDublinCore: addBelow(state.currentDublinCore, field, index, value),
        }
    } else if (type === "DELETE_VALUE") {
        const { field, index } = payload as { field: string, index: number };
        return {
            ...state,
            keyCounter: incrementKeyCounter(state.keyCounter, field),
            currentDublinCore: deleteValue(state.currentDublinCore, field, index),
        }
    } else if (type === "REPLACE_VALUE") {
        const { field, index, value } = payload as { field: string, index: number, value: string };
        return {
            ...state,
            // We don't need to refresh the key counter in this scenario, because the number of
            // fields is not changing.
            currentDublinCore: replaceValue(state.currentDublinCore, field, index, value),
        }
    } else if(Object.keys(reducerMapping).includes(type)) {
        return {
            ...state,
            [reducerMapping[type]]: payload
        };
    }
    console.error(`processMetadata action type: ${type} does not exist`);
    return state;
};

export const DublinCoreMetadataContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(dublinCoreMetadataReducer, dublinCoreContextParams);
    const value = { state, dispatch };
    return <DublinCoreMetadataContext.Provider value={value}>{children}</DublinCoreMetadataContext.Provider>;
};

export const useDublinCoreMetadataContext = () => {
    const {
        state,
        dispatch,
    } = useContext(DublinCoreMetadataContext) as { state: DublinCoreMetadata, dispatch: (params: unknown) => void };

    const setCurrentDublinCore = (dc: Record<string, Array<string>>) => {
        dispatch({
            type: "SET_CURRENT_DUBLIN_CORE",
            payload: dc
        });
    };

    const addValueAbove = (field: string, index: number, value: string): void => {
        dispatch({
            type: "ADD_VALUE_ABOVE",
            payload: { field, index, value }
        });
    }

    const addValueBelow = (field: string, index: number, value: string): void => {
        dispatch({
            type: "ADD_VALUE_BELOW",
            payload: { field, index, value }
        });
    }

    const deleteValue = (field: string, index: number): void => {
        dispatch({
            type: "DELETE_VALUE",
            payload: { field, index }
        });
    }

    const replaceValue = (field: string, index: number, value: string): void => {
        dispatch({
            type: "REPLACE_VALUE",
            payload: { field, index, value }
        });
    }

    return {
        state,
        action: {
            addValueAbove,
            addValueBelow,
            deleteValue,
            replaceValue,
            setCurrentDublinCore,
        },
    };
};

DublinCoreMetadataContextProvider.propTypes = {
    children: PropTypes.node,
};

export default { DublinCoreMetadataContextProvider, useDublinCoreMetadataContext };
