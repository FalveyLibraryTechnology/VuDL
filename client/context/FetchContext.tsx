import React, { createContext, useContext, useReducer } from "react";
import { authApiUrl, ingestApiUrl, loginUrl } from "../util/routes";
import PropTypes from "prop-types";

/**
 * Pass a shared entity to react components,
 * specifically a way to make api requests.
 */
const fetchContextParams = {
    token: typeof window !== 'undefined' ? sessionStorage.getItem("token") : null,
};
const FetchContext = createContext({});

/**
 * Update the shared states of react components.
 */
const fetchReducer = (state, { type, payload }) => {
    switch (type) {
        case "UPDATE_TOKEN":
            sessionStorage.setItem("token", payload);
            return {
                ...state,
                token: payload,
            };
        default:
            console.error(`fetch action type: ${type} does not exist`);
            return state;
    }
};

export const FetchContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(fetchReducer, fetchContextParams);
    const value = { state, dispatch };
    return <FetchContext.Provider value={value}>{children}</FetchContext.Provider>;
};

export const useFetchContext = () => {
    const {
        state: { token },
        dispatch,
    } = useContext(FetchContext);

    /**
     * Get default fetch parameters. If params or headers exist,
     * overload the associated keys.
     */
    const fetchParams = (params = {}, headers = {}) => ({
        credentials: "include",
        method: "GET",
        mode: "cors",
        ...params,
        headers,
    });

    /**
     * Query a fresh token to authorize user to make api requests.
     */
    const refreshToken = async () => {
        try {
            const response = await fetch(`${authApiUrl}/token/mint`, fetchParams());
            if (response.ok) {
                const token = await response.json();
                dispatch({
                    type: "UPDATE_TOKEN",
                    payload: token,
                });
                return token;
            }
        } catch (error) {
            console.error(error);
        }
        window.location.href = `${loginUrl}?referer=${encodeURIComponent(window.location.href)}`;
        return false;
    };

    /**
     * Return a customizable response when making a request.
     * If first response return a 401 unauthorized status,
     * attempt to refresh token and call the request again.
     * @param {string} url - The http url of the request
     * @param {Object} params - The request parameters
     * @param {Object} headers - The request parameter headers
     */
    const makeRequest = async (url, params = {}, headers = {}) => {
        const response = await fetch(
            url,
            fetchParams(params, {
                ...headers,
                Authorization: `Token ${token}`,
            })
        );
        if (response.status == 401) {
            const token = await refreshToken();
            return token
                ? await fetch(
                      url,
                      fetchParams(params, {
                          ...headers,
                          Authorization: `Token ${token}`,
                      })
                  )
                : false;
        }
        return response;
    };

    /**
     * Return a json response when making a request
     * @param {string} url - The http url of the request
     * @param {Object} params - The request parameters
     * @param {Object} headers - The request parameter headers
     */
    const fetchJSON = async (url = ingestApiUrl, params = {}, headers = {}) => {
        const response = await makeRequest(url, params, headers);
        if (response?.ok) {
            return await response.json();
        }
        throw new Error(response?.statusText ?? "Could not make request");
    };

    /**
     * Return a text response when making a request
     * @param {string} url - The http url of the request
     * @param {Object} params - The request parameters
     * @param {Object} headers - The request parameter headers
     */
    const fetchText = async (url = ingestApiUrl, params = {}, headers = {}) => {
        const response = await makeRequest(url, params, headers);
        if (response.ok) {
            return await response.text();
        }
        throw new Error(response.statusText);
    };

    return {
        state: {
            token,
        },
        action: {
            fetchJSON,
            fetchText,
            makeRequest,
        },
    };
};

FetchContextProvider.propTypes = {
    children: PropTypes.node,
};

export default { FetchContextProvider, useFetchContext };
