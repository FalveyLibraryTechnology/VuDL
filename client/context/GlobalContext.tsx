import React, { createContext, useContext, useReducer } from "react";

export enum ModalName {
    DatastreamModal = "datastream",
    ParentModal = "parents",
    StateModal = "state",
};

interface SnackbarState {
    open: boolean,
    message: string,
    severity: string
}

interface GlobalState {
    snackbarState: SnackbarState;
}

/**
 * Pass a shared entity to react components,
 * specifically a way to make api requests.
 */
const globalContextParams: GlobalState = {
    // Modal control
    openModalState: new Set(),
    // Snackbar
    snackbarState: {
        open: false,
        message: "",
        severity: "info",
    },
};

const reducerMapping: Record<string, string> = {
    // Snackbar
    SET_SNACKBAR_STATE: "snackbarState",
};

/**
 * Update the shared states of react components.
 */
const globalReducer = (state: GlobalState, { type, payload }: { type: string, payload: unknown}) => {
    if (type == "OPEN_MODAL") {
        state.openModalState.add(payload);
        return state;
    }
    if (type == "CLOSE_MODAL") {
        state.openModalState.remove(payload);
        return state;
    }

    if (Object.keys(reducerMapping).includes(type)){
        return {
            ...state,
            [reducerMapping[type]]: payload
        };
    } else {
        console.error(`global action type: ${type} does not exist`);
        return state;
    }
};

const GlobalContext = createContext({});

export const GlobalContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(globalReducer, globalContextParams);
    const value = { state, dispatch };
    return <GlobalContext.Provider value={value}>{children}</GlobalContext.Provider>;
};

export const useGlobalContext = () => {
    const {
        state: {
            // Modal control
            openModalState,
            // Snackbar
            snackbarState,
        },
        dispatch,
    } = useContext(GlobalContext);

    // Modal control

    const isModalOpen = (modal: ModalName) => {
        return openModalState.has(modal);
    };
    const openModal = (modal: ModalName) => {
        dispatch({
            type: "OPEN_MODAL",
            payload: modal
        });
    };
    const closeModal = (modal: ModalName) => {
        dispatch({
            type: "CLOSE_MODAL",
            payload: modal
        });
    };
    const toggleModal = (modal: ModalName) => {
        if (isModalOpen(modal)) {
            closeModal(modal);
        } else {
            openModal(modal);
        }
    };

    // Snackbar

    const setSnackbarState = (snackbarState: SnackbarState) => {
        dispatch({
            type: "SET_SNACKBAR_STATE",
            payload: snackbarState
        });
    };

    return {
        state: {
            // Snackbar
            snackbarState,
        },
        action: {
            // Modal control
            isModalOpen,
            openModal,
            closeModal,
            toggleModal,
            // Snackbar
            setSnackbarState,
        },
    };
}

export default {
    GlobalContextProvider,
    useGlobalContext
}
