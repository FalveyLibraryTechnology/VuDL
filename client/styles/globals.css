:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
