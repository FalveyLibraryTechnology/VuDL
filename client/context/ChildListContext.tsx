import React, { createContext, useContext, useReducer } from "react";
import { getObjectChildrenUrl } from "../util/routes";
import { useFetchContext } from "./FetchContext";

interface Children {
    numFound?: number;
    start?: number;
    docs?: Record<string, string>[];
}

interface ChildListState {
    childStorage: Record<string, Children>;
}

const childListContextParams: ChildListState = {
    childStorage: {},
};

const ChildListContext = createContext({});

const reducerMapping: Record<string, string> = {
};

/**
 * Update the shared states of react components.
 */
const childListReducer = (state: ChildListState, { type, payload }: { type: string, payload: unknown}) => {
    if (type === "ADD_TO_CHILD_STORAGE") {
        const { key, children } = payload;
        const childStorage = {
            ...state.childStorage,
        };
        childStorage[key] = children;
        return {
            ...state,
            childStorage
        };
    } else if (Object.keys(reducerMapping).includes(type)){
        return {
            ...state,
            [reducerMapping[type]]: payload
        };
    } else {
        console.error(`fetch action type: ${type} does not exist`);
        return state;
    }
};

export const ChildListContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(childListReducer, childListContextParams);
    const value = { state, dispatch };
    return <ChildListContext.Provider value={value}>{children}</ChildListContext.Provider>;
};

export const useChildListContext = () => {
    const {
        action: {
            fetchJSON
        }
    }= useFetchContext();
    const {
        state: {
            childStorage
        },
        dispatch,
    } = useContext(ChildListContext);

    const addToChildStorage = (key: string, children: Children) => {
        dispatch({
            type: "ADD_TO_CHILD_STORAGE",
            payload: { key, children },
        });
    };

    const getChildStorageKey = (pid: string, page: number, pageSize: number): string => {
        return `${pid}_${page}_${pageSize}`;
    }

    const loadChildrenIntoStorage = async (pid: string, page: number, pageSize: number) => {
        const key = getChildStorageKey(pid, page, pageSize);
        const url = getObjectChildrenUrl(pid, (page - 1) * pageSize, pageSize);
        try {
            addToChildStorage(key, await fetchJSON(url));
        } catch (e) {
            console.error("Problem fetching tree data from " + url);
        }
    }

    return {
        state: {
            childStorage,
        },
        action: {
            getChildStorageKey,
            loadChildrenIntoStorage,
        },
    };
}

export default {
    ChildListContextProvider,
    useChildListContext
}