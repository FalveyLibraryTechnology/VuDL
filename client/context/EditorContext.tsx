import React, { createContext, useContext, useReducer } from "react";
import { editObjectCatalogUrl, getObjectChildrenUrl, getObjectDetailsUrl } from "../util/routes";
import { useFetchContext } from "./FetchContext";
import { extractFirstMetadataValue as utilExtractFirstMetadataValue } from "../util/metadata";

export interface ObjectDetails {
    fedoraDatastreams: Array<string>;
    metadata: Record<string, Array<string>>;
    models: Array<string>;
    pid: string;
    sortOn: string;
    state: string;
};

interface ChildrenResultPage {
    numFound?: number;
    start?: number;
    docs?: Record<string, string|string[]>[];
}

interface SnackbarState {
    open: boolean,
    message: string,
    severity: string
}

export interface FedoraDatastream {
    mimetype?: {
        allowedType: string;
        allowedSubtypes: string;
    };
}

export interface FedoraModel {
    datastreams?: Record<string, FedoraDatastream>;
}

export interface License {
    name: string;
    uri: string;
}

interface EditorState {
    modelsCatalog: Record<string, FedoraModel>;
    licensesCatalog: Record<string, License>;
    currentPid: string | null;
    activeDatastream: string | null;
    isDatastreamModalOpen: boolean;
    isStateModalOpen: boolean;
    datastreamModalState: string | null;
    stateModalActivePid: string | null;
    snackbarState: SnackbarState;
    objectDetailsStorage: Record<string, ObjectDetails>;
    childListStorage: Record<string, ChildrenResultPage>;
}

/**
 * Pass a shared entity to react components,
 * specifically a way to make api requests.
 */
const editorContextParams: EditorState = {
    modelsCatalog: {},
    licensesCatalog: {},
    currentPid: null,
    activeDatastream: null,
    isDatastreamModalOpen: false,
    isStateModalOpen: false,
    datastreamModalState: null,
    stateModalActivePid: null,
    snackbarState: {
        open: false,
        message: "",
        severity: "info"
    },
    objectDetailsStorage: {},
    childListStorage: {},
};

export const DatastreamModalStates = {
    UPLOAD: "Upload",
    VIEW: "View",
    METADATA: "Metadata",
    DOWNLOAD: "Download",
    DELETE: "Delete"
};


const EditorContext = createContext({});

const reducerMapping: Record<string, string> = {
    SET_LICENSES_CATALOG: "licensesCatalog",
    SET_MODELS_CATALOG: "modelsCatalog",
    SET_CURRENT_PID: "currentPid",
    SET_ACTIVE_DATASTREAM: "activeDatastream",
    SET_IS_DATASTREAM_MODAL_OPEN: "isDatastreamModalOpen",
    SET_IS_STATE_MODAL_OPEN: "isStateModalOpen",
    SET_DATASTREAM_MODAL_STATE: "datastreamModalState",
    SET_STATE_MODAL_ACTIVE_PID: "stateModalActivePid",
    SET_SNACKBAR_STATE: "snackbarState",
};

/**
 * Update the shared states of react components.
 */
const editorReducer = (state: EditorState, { type, payload }: { type: string, payload: SnackbarState | unknown}) => {
    if (type === "ADD_TO_OBJECT_DETAILS_STORAGE") {
        const { key, details } = payload as { key: string; details: ObjectDetails };
        const objectDetailsStorage = {
            ...state.objectDetailsStorage,
        };
        objectDetailsStorage[key] = details;
        return {
            ...state,
            objectDetailsStorage
        };
    } else if (type === "REMOVE_FROM_OBJECT_DETAILS_STORAGE") {
        const { key } = payload as { key: string };
        const objectDetailsStorage = {
            ...state.objectDetailsStorage,
        };
        delete objectDetailsStorage[key];
        return {
            ...state,
            objectDetailsStorage
        };
    } else if (type === "ADD_TO_CHILD_LIST_STORAGE") {
        const { key, children } = payload as { key: string; children: ChildrenResultPage };
        const childListStorage = {
            ...state.childListStorage,
        };
        childListStorage[key] = children;
        return {
            ...state,
            childListStorage
        };
    } else if(Object.keys(reducerMapping).includes(type)){
        return {
            ...state,
            [reducerMapping[type]]: payload
        };
    } else {
        console.error(`fetch action type: ${type} does not exist`);
        return state;
    }
};

export const EditorContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(editorReducer, editorContextParams);
    const value = { state, dispatch };
    return <EditorContext.Provider value={value}>{children}</EditorContext.Provider>;
};

export const useEditorContext = () => {
    const {
        action: {
            fetchJSON
        }
    }= useFetchContext();
    const {
        state: {
            currentPid,
            activeDatastream,
            isDatastreamModalOpen,
            isStateModalOpen,
            datastreamModalState,
            stateModalActivePid,
            licensesCatalog,
            modelsCatalog,
            snackbarState,
            objectDetailsStorage,
            childListStorage,
        },
        dispatch,
    } = useContext(EditorContext);

    const currentDatastreams = objectDetailsStorage?.[currentPid]?.datastreams ?? [];
    const currentModels = objectDetailsStorage?.[currentPid]?.models ?? [];
    const modelsDatastreams = currentModels.reduce((acc: Array<string>, model: string) => {
        const name = model.split(":")?.[1];
        return [
            ...acc,
            ...(name && modelsCatalog.hasOwnProperty(name) ? Object.keys(modelsCatalog[name].datastreams) : [])
        ];
    }, []).map((stream: string) => {
        return {
            disabled: !currentDatastreams.includes(stream),
            stream
        };
    });

    const addToObjectDetailsStorage = (key: string, details: ObjectDetails) => {
        console.log("add");
        dispatch({
            type: "ADD_TO_OBJECT_DETAILS_STORAGE",
            payload: { key, details },
        });
    };

    const removeFromObjectDetailsStorage = (key: string) => {
        console.log("remove");
        dispatch({
            type: "REMOVE_FROM_OBJECT_DETAILS_STORAGE",
            payload: { key },
        });
    };

    const addToChildListStorage = (key: string, children: ChildrenResultPage) => {
        dispatch({
            type: "ADD_TO_CHILD_LIST_STORAGE",
            payload: { key, children },
        });
    };

    const getChildListStorageKey = (pid: string, page: number, pageSize: number): string => {
        return `${pid}_${page}_${pageSize}`;
    };

    const loadObjectDetailsIntoStorage = async (pid: string) => {
        console.log("load");
        // Ignore null values:
        if (pid === null) {
            return;
        }
        const url = getObjectDetailsUrl(pid);
        try {
            addToObjectDetailsStorage(pid, await fetchJSON(url));
        } catch (e) {
            console.error("Problem fetching details from " + url);
        }
    };

    const loadChildrenIntoStorage = async (pid: string, page: number, pageSize: number) => {
        const key = getChildListStorageKey(pid, page, pageSize);
        const url = getObjectChildrenUrl(pid, (page - 1) * pageSize, pageSize);
        try {
            addToChildListStorage(key, await fetchJSON(url));
        } catch (e) {
            console.error("Problem fetching tree data from " + url);
        }
    };

    const setModelsCatalog = (modelsCatalog: Record<string, FedoraModel>) => {
        dispatch({
            type: "SET_MODELS_CATALOG",
            payload: modelsCatalog
        });
    };

    const setLicensesCatalog = (licensesCatalog: Record<string, License>) => {
        dispatch({
            type: "SET_LICENSES_CATALOG",
            payload: licensesCatalog
        });
    };

    const setCurrentPid = (pid: string) => {
        dispatch({
            type: "SET_CURRENT_PID",
            payload: pid
        });
    };

    const toggleDatastreamModal = () => {
        dispatch({
            type: "SET_IS_DATASTREAM_MODAL_OPEN",
            payload: !isDatastreamModalOpen
        });
    };

    const toggleStateModal = () => {
        dispatch({
            type: "SET_IS_STATE_MODAL_OPEN",
            payload: !isStateModalOpen
        });
    };

    const setDatastreamModalState = (datastreamModalState: boolean) => {
        dispatch({
            type: "SET_DATASTREAM_MODAL_STATE",
            payload: datastreamModalState
        });
    };

    const setStateModalActivePid = (pid: string) => {
        dispatch({
            type: "SET_STATE_MODAL_ACTIVE_PID",
            payload: pid
        });
    };

    const setActiveDatastream = (datastream: string) => {
        dispatch({
            type: "SET_ACTIVE_DATASTREAM",
            payload: datastream
        })
    };

    const setSnackbarState = (snackbarState: SnackbarState) => {
        dispatch({
            type: "SET_SNACKBAR_STATE",
            payload: snackbarState
        });
    };

    const datastreamsCatalog = Object.values(modelsCatalog).reduce((acc: Record<string, FedoraDatastream>, model) => {
        return {
            ...acc,
            ...(model as FedoraModel).datastreams
        };
    }, {});

    const initializeCatalog = async () => {
        try {
            const response = await fetchJSON(editObjectCatalogUrl);
            setModelsCatalog(response.models || {});
            setLicensesCatalog(response.licenses || {});
        } catch(err) {
            console.error(`Problem fetching object catalog from ${editObjectCatalogUrl}`);
        }
    };

    const loadCurrentObjectDetails = async () => {
        return await loadObjectDetailsIntoStorage(currentPid);
    };

    const extractFirstMetadataValue = function (field: string, defaultValue: string): string {
        const currentMetadata = objectDetailsStorage?.[currentPid]?.metadata ?? {};
        return utilExtractFirstMetadataValue(currentMetadata, field, defaultValue);
    }

    return {
        state: {
            currentPid,
            currentDatastreams,
            activeDatastream,
            isDatastreamModalOpen,
            isStateModalOpen,
            datastreamModalState,
            stateModalActivePid,
            datastreamsCatalog,
            modelsDatastreams,
            modelsCatalog,
            licensesCatalog,
            snackbarState,
            objectDetailsStorage,
            childListStorage,
        },
        action: {
            initializeCatalog,
            setCurrentPid,
            loadCurrentObjectDetails,
            setActiveDatastream,
            setDatastreamModalState,
            setStateModalActivePid,
            toggleDatastreamModal,
            toggleStateModal,
            setSnackbarState,
            extractFirstMetadataValue,
            getChildListStorageKey,
            loadObjectDetailsIntoStorage,
            loadChildrenIntoStorage,
            removeFromObjectDetailsStorage,
        },
    };
}

export default {
    EditorContextProvider,
    DatastreamModalStates,
    useEditorContext
}