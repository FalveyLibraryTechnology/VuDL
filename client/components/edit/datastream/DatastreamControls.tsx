import React from "react";
import PropTypes from "prop-types";
import { Box, ButtonGroup } from "@mui/material";
import { DatastreamModalStates } from "../../../context/EditorContext";
import DatastreamControlButton from "./DatastreamControlButton";

const DatastreamControls = ({ datastream, disabled }) => {
    return (
        <Box className="datastreamControls">
            <ButtonGroup variant="text">
                {Object.values(DatastreamModalStates).map((modalState, index) => {
                    return (
                        <DatastreamControlButton
                            modalState={modalState}
                            disabled={disabled}
                            datastream={datastream}
                            key={index}
                        />
                    );
                })}
            </ButtonGroup>
        </Box>
    );
};
DatastreamControls.propTypes = {
    datastream: PropTypes.string,
    disabled: PropTypes.bool,
};
export default DatastreamControls;
