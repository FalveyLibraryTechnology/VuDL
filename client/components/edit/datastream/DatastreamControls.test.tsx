import React from "react";
import { describe, expect, it, jest } from "@jest/globals";
import renderer from "react-test-renderer";
import { DatastreamModalStates } from "../../../context/EditorContext";
import DatastreamControls from "./DatastreamControls";

const mockDatastreamControlButton = jest.fn();
jest.mock("./DatastreamControlButton", () => (props) => {
    mockDatastreamControlButton(props);
    return "DatastreamControlButton: " + JSON.stringify(props);
});

describe("DatastreamControls", () => {
    it("renders", () => {
        const datastream = "test1";
        const tree = renderer.create(<DatastreamControls datastream={datastream} disabled={false} />).toJSON();
        expect(tree).toMatchSnapshot();
        expect(mockDatastreamControlButton).toHaveBeenCalledWith({
            modalState: DatastreamModalStates.UPLOAD,
            datastream,
            disabled: false,
        });
    });
});
