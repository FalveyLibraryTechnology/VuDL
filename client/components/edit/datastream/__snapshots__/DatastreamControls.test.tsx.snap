// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatastreamControls renders 1`] = `
<div
  class="datastreamControls MuiBox-root css-0"
>
  <div
    class="MuiButtonGroup-root MuiButtonGroup-text css-1or64il-MuiButtonGroup-root"
    role="group"
  >
    DatastreamControlButton
  </div>
</div>
`;
